### PR TITLE
JetBrains: add up/down shortcut in Cody chat input to iterate through previous messages

### PR DESCRIPTION
**Pull request by Sa1to**
It helps #54933 
## Test plan
* Open new cody chat window
* Send message 1
* Send message 2
* Press ⬆️, message 2 should appear in the prompt
* Press ⬆️, message 1 should appear in the prompt
* Press ⬇️, message 2 should again appear in the prompt
* Modify message 2 and send
* Press ⬆️, modified message 2 should appear 

https://github.com/sourcegraph/sourcegraph/assets/9321940/5aefba04-f7eb-441c-ae23-9a4169ff3be8

### CODE REVIEW

File: client/jetbrains/src/main/kotlin/com/sourcegraph/cody/chat/CodyChatMessageHistory.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            promptInput.text = pop
+            currentValue = pop
+        }
+    }
+
+    /**
+     * When new message is sent it is pushing all messages
+     *  from lower stack to upper stack and at the end pushes new message
+     */
+    fun messageSent(promptInput: JBTextArea) {
```
Comment (mkondratek) [inline client/jetbrains/src/main/kotlin/com/sourcegraph/cody/chat/CodyChatMessageHistory.kt:35]: I am just thinking... 🤔 what is the expected behavior in this case:

1. upper stack contains a former prompt e.g. `find method int foo()`
2. prompt input is modified e.g. `find method void bar()`
3. ⬆️ 

I guess (and iirc thats how terminals control+r works) it should move back to former prompt (`find method int foo()`) but keeps the inserted input in the case of hitting ⬇️.

--- 

The other issue I see is caret navigation. I assume one can write a multiline prompt. e.g.

```
hello cody, you are very helpful, I have a request for you, 
find me a method called "buz()" that return "String".👇 
```

In some text input boxes hitting ⬆️ moves caret to the upper line, so the result should be smth like this:

```
hello cody, you are very helpful, I have a request for 👇you, 
find me a method called "buz()" that return "String". 
```

Idk what is expected in this case.

**Review by olafurpg — CHANGES_REQUESTED** [review]
I just tried this locally and it currently doesn't match my expected behavior:

Pressing down/up to cycle through history must only happen if the message is empty or if the last keyboard action was an up/down keystroke to iterate through history. Currently, I can press up, type a custom message, press down and my custom message gets lost! We must never loose the contents of a custom message


CleanShot 2023-09-08 at 11 54 12


If I start editing a message or move the cursor to the left/right, the behavior of up/down must get reset. The only way to start going up/down through history again is by deleting the entire message so that the input box is empty again. The first time I press up after clearing the input box, the counter should also be reset, so that pressing up should display the most recent message (not the value of the counter from an old up/down action).

Does that makes sense?

File: client/jetbrains/src/main/kotlin/com/sourcegraph/cody/chat/CodyChatMessageHistory.kt (hunk 1 above)
Comment (Sa1to) [reply client/jetbrains/src/main/kotlin/com/sourcegraph/cody/chat/CodyChatMessageHistory.kt:35]: 1. Good point, my last commit should fix it
2. I wasn't able to reproduce it... can you send a screen recording or something?

**Comment by olafurpg** [comment]
@Sa1to The following behavior is still surprising

- Start fresh chat
- Type "Hello"
- Press `Down`
- Input box is empty, I expected it to have no effect to press down

Similar for `Up`, I expect the arrow to move to the first character of the input box.

**Comment by Sa1to** [comment]
> In general, my expectation is that up/down history should never get triggered except in the scenario where I begin from an empty input box and _exclusively_ press up/down. As soon as I press any other character, I must empty the input box to be able to use up/down to go through history again.

@olafurpg I have added behaviour like in vs code plugin. When there is text in the prompt, pressing ⬆️  jumps to the beginning of the prompt, if you are at the beginning it puts last sent message into the prompt.

**Comment by szymonprz** [comment]
>Press up twice
Expected behavior: cursor to move at the start of the message, stays there on the second up
Actual behavior: cursor moves first to the start of the message, and changes history entry the second time I press up

Why do we want such behaviour? I thought that when I move through the history our cursor will stay at the end of the message so I can add/edit something more easily. I think that it's more probable that I need to edit something at the end than at the beginning of the message so now every when I choose something from history I must go to the end of the message to add something, this can be really annoying.

My idea of this is:
- You can only go through the history if the chat is empty or you are going through the history and the cursor is always at the end of the message from the history
- You can't go through the history if you edited the chat message and your cursor stays in place

wdyt?

**Comment by olafurpg** [comment]
@szymonprz the cursor should only move to the start of the message if you have reached the *first* message. I agree that the current behavior in latest commit is annoying. The cursor should stay at the end of the message while you press up until you reach the first message, in which pressing `up` has the default behavior.

The short summary of the behavior of `up`/`down` arrow keys:

- By default, up/down should behave exactly like normal
- If the input box is empty, you can press up to go to the previous message. Doing this enters "history mode" where up/down arrow keys move through the history instead of having default behavior
- You exit "history mode" when you make a non-up/down arrow keystroke OR if you reach the first/last message.

**Comment by Sa1to** [comment]
> @szymonprz the cursor should only move to the start of the message if you have reached the _first_ message. I agree that the current behavior in latest commit is annoying. The cursor should stay at the end of the message while you press up until you reach the first message, in which pressing `up` has the default behavior.
> 
> The short summary of the behavior of `up`/`down` arrow keys:
> 
> * By default, up/down should behave exactly like normal
> * If the input box is empty, you can press up to go to the previous message. Doing this enters "history mode" where up/down arrow keys move through the history instead of having default behavior
> * You exit "history mode" when you make a non-up/down arrow keystroke OR if you reach the first/last message.

Thats fine by me, just FYI if we do it like so it won't be possible to modify one of the history entires and send it as a new message.

File: client/jetbrains/src/main/java/com/sourcegraph/cody/CodyToolWindowContent.java [hunk 2]
```diff
@@ -115,7 +119,40 @@ public CodyToolWindowContent(@NotNull Project project) {
...
+    AnAction upperMessageAction =
+        new DumbAwareAction() {
+          @Override
+          public void actionPerformed(@NotNull AnActionEvent e) {
+            if (!chatMessageHistory.getUpperStack().isEmpty()
+                && (promptInput.getText().isEmpty()
+                    || promptInput.getText().equals(chatMessageHistory.getCurrentValue()))) {
+              chatMessageHistory.popUpperMessage(promptInput);
+            } else {
+              promptInput.setCaretPosition(0);
```
Comment (olafurpg) [inline client/jetbrains/src/main/java/com/sourcegraph/cody/CodyToolWindowContent.java:138]: Moving the cursor to caret position 0 does not mirror the default behavior of `UP` in a text area. The default behavior of `UP` is to move to the line above. Is there a way to trigger the default `UP` behavior?

**Review by olafurpg — CHANGES_REQUESTED** [review]
I tried this locally and the history behavior itself works great. I have one last blocking comment wrt. to the fallback behavior of UP/DOWN

**Comment by Sa1to** [comment]
This shoudl be fixed now:

https://github.com/sourcegraph/sourcegraph/assets/9321940/044e0857-7f01-4078-901a-38482bffeaf2

**Review by olafurpg — CHANGES_REQUESTED** [review]
I just tried this locally and observed one surprising behavior

- Send multiline message
- Press up multiple times until I'm at the first character of the first messag
- Press down. I expect `Down` to behave like usual but instead the message disappears (because it's going down in message history)

CleanShot 2023-09-25 at 18 46 44

**Comment by olafurpg** [comment]
The latest behavior still has one surprising behavior

- Send message `a`
- Send multiline message `b\nc\nd`
- Press up twice
- I expected the cursor to be at the first message `a` but the second up keystroke moved to after `c`

CleanShot 2023-10-02 at 18 29 20

File: client/jetbrains/src/main/java/com/sourcegraph/cody/CodyToolWindowContent.java [hunk 3]
```diff
@@ -117,16 +125,99 @@ public CodyToolWindowContent(@NotNull Project project) {
...
+                  int newCaretPosition =
+                      Math.min(maxPreviousLineOffset - 1, previousLineStart + positionInLine);
+                  promptInput.setCaretPosition(newCaretPosition);
+                }
+                // When on the first line, move caret to the beginning of the line.
+                else {
+                  promptInput.setCaretPosition(0);
+                }
+              } catch (BadLocationException ex) {
+                throw new RuntimeException(ex);
```
Comment (mkondratek) [inline client/jetbrains/src/main/java/com/sourcegraph/cody/CodyToolWindowContent.java:159]: are we okay with a throw like this? 😕

**Review by mkondratek — COMMENTED** [review]
I tested it locally - it worked for me

**Comment by olafurpg** [comment]
Closing since this is outdated.

**Pull request closed without merging**